[ZEPPELIN-1334] Environment variable defined in interpreter setting doesn't take effect

zjffdu opened the pull request:
### What is this PR for?
I define SPAKR_HOME in interpreter setting, but it doesn't take effect. This PR is for bring back the environment variable defined in interpreter setting.  The root cause is that we reset the env after creating RemoteInterpreter.
```
         new RemoteInterpreter(property, noteId, className, conf.getInterpreterRemoteRunnerPath(),
             interpreterPath, localRepoPath, connectTimeout, maxPoolSize,
             remoteInterpreterProcessListener, appEventListener);
    remoteInterpreter.setEnv(env);
```


### What type of PR is it?
[Bug Fix]

### Todos
* [ ] - Task

### What is the Jira issue?
* https://issues.apache.org/jira/browse/ZEPPELIN-1334

### How should this be tested?
Tested manually.  Create 2 spark interpreter setting, one for spark1 another is for spark2. And define SPARK_HOME for each interpreter. Then I can run both spark1 and spark2 in one zeppelin instance. 

### Screenshots (if appropriate)
![image](https://cloud.githubusercontent.com/assets/164491/17696073/b64b1014-63de-11e6-88ab-d26b1c2fa75c.png)

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

zjffdu commented:
@Leemoonsoo Please help to review.

Leemoonsoo commented:
Thanks @zjffdu for the fix. Let me take a look.

bzz commented:
@zjffdu thank you for fixing it!

After the review we better merge this to both, _master_ and _branch-0.6_

zjffdu commented:
@bzz Sure, let me know if I need to create another PR for 0.6

Leemoonsoo commented:
@zjffdu Tested and It looks good to me.

There is an unittest for set environment variable for remote interpreter at [RemoteInterpreterTest.java](https://github.com/apache/zeppelin/blob/master/zeppelin-interpreter/src/test/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterTest.java#L721). However, the unit test is not testing the case this fix addresses.

It's not going to be a blocker but I think it'll be really beneficial to have a unittest that covers this fix, somewhere in [InterpreterFactoryTest.java](https://github.com/apache/zeppelin/blob/master/zeppelin-interpreter/src/test/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterTest.java). @zjffdu Is it going to be difficult to add the test?  Let me know if you need any help!

zjffdu commented:
@Leemoonsoo  I have added unit test, please help review. failed test seems irrelevant.

jongyoul commented on zeppelin-zengine/src/test/java/org/apache/zeppelin/interpreter/InterpreterFactoryTest.java:
```diff
@@ -18,10 +18,7 @@
-import java.util.LinkedList;
-import java.util.List;
-import java.util.Properties;
+import java.util.*;
```
Import packages with wildcard is not recommended. Could you please check it?

bzz replied:
As per project style guilde https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports

bzz commented:
@zjffdu Looks great to me, modulo minor style issue above, thank you and :+1: for tests!

One more thing - it's a bit strange ZEPPELIN-1334 has fix-for version ONLY 0.6.2 which implies merging only to _branch-0.6_ but this PR is to _master_.

Let's merge it, as soon as those 2 issues are addressed!

bzz commented:
All 3 CI failing profiles: [Spark 2.0](https://s3.amazonaws.com/archive.travis-ci.org/jobs/153792236/log.txt) and 1.6 are very consistent 
````
[INFO] Zeppelin: Zengine .................................. FAILURE [ 44.356 s]

Failed tests: 
  InterpreterFactoryTest.testRemoteRepl:118 expected:<VALUE_1> but was:<null>


Tests run: 6, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.108 sec <<< FAILURE! - in org.apache.zeppelin.interpreter.InterpreterFactoryTest
testRemoteRepl(org.apache.zeppelin.interpreter.InterpreterFactoryTest)  Time elapsed: 0.012 sec  <<< FAILURE!
java.lang.AssertionError: expected:<VALUE_1> but was:<null>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:743)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.apache.zeppelin.interpreter.InterpreterFactoryTest.testRemoteRepl(InterpreterFactoryTest.java:118)
```

@zjffdu  could you please take a look and make sure it's not related to the changes?

zjffdu commented:
Thanks @bzz, I have fixed the code style issue. Another remaining thing is that the unit test I added is failed when running it using maven with other unit test. But it succeeded when I ran it individually. this is very weird.

bzz commented:
@zjffdu so there must be some state that persists between the runs. 

@Leemoonsoo do you have any idea what that state might be that affects `InterpreterFactoryTest `? There seems to be nothing like `@BeforeClass`, etc.

Shall we try re-triggering CI one more time?

bzz commented:
Very similar test failure as we discussed above happened in un-related #1363 [here](https://s3.amazonaws.com/archive.travis-ci.org/jobs/155836675/log.txt)

```

Failed tests: 
  InterpreterFactoryTest.testRemoteRepl:123 expected:<VALUE_1> but was:<null>


Tests run: 6, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.146 sec <<< FAILURE! - in org.apache.zeppelin.interpreter.InterpreterFactoryTest
testRemoteRepl(org.apache.zeppelin.interpreter.InterpreterFactoryTest)  Time elapsed: 0.032 sec  <<< FAILURE!
java.lang.AssertionError: expected:<VALUE_1> but was:<null>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:743)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.apache.zeppelin.interpreter.InterpreterFactoryTest.testRemoteRepl(InterpreterFactoryTest.java:123)

```

zjffdu commented:
PR rebased.

zjffdu commented:
@Leemoonsoo @bzz I just fix the unit test. The root cause is that `HeliumApplicationFactoryTest` will also register these interpreters. And all these unit tests are in the same JVM. So I change `pom.xml` to fork new jvm for each unit test.

Leemoonsoo commented:
Great work @zjffdu!

Looks good to me!

Leemoonsoo commented:
Merge it to master and branch-0.6 if there're no more discussions.
